src/liballoc/string.rs

```diff
@@ -1982,6 +1982,124 @@ impl Add<&str> for String {
     }
 }
 
+#[allow(missing_docs)]
+#[stable(feature = "add_string_and_dbl_ref_str", since = "1.41.0")]
+impl Add<&&str> for String {
+    type Output = String;
+
+    #[inline]
+    fn add(mut self, other: &&str) -> String {
+        self.push_str(other);
+        self
+    }
+}
+
+/// Implements the `+` operator for concatenating two `String`s.
+///
+/// This consumes the `String` on the left-hand side and re-uses its buffer (growing it if
+/// necessary). This is done to avoid allocating a new `String` and copying the entire contents on
+/// every operation, which would lead to `O(n^2)` running time when building an `n`-byte string by
+/// repeated concatenation.
+///
+/// The string on the right-hand side is only borrowed; its contents are copied into the returned
+/// `String`.
+///
+/// # Examples
+///
+/// Concatenating two `String`s takes the first by value and borrows the second:
+///
+/// ```
+/// let a = String::from("hello");
+/// let b = String::from(" world");
+/// let c = a + &b;
+/// // `a` is moved and can no longer be used here.
+/// ```
+///
+/// If you want to keep using the first `String`, you can clone it and append to the clone instead:
+///
+/// ```
+/// let a = String::from("hello");
+/// let b = String::from(" world");
+/// let c = a.clone() + &b;
+/// // `a` is still valid here.
+/// ```
+///
+/// Concatenating `&str` slices can be done by converting the first to a `String`:
+///
+/// ```
+/// let a = "hello";
+/// let b = " world";
+/// let c = a.to_string() + b;
+/// ```
```

Comment:
This whole comment is just a copy from the `Add<&str>` impl, right? I would probably:

- Add the information about the "problem" we are facing to the old comment for `Add<&str>`, i.e.: explain that we also had to add strange impls like `Add<&String>` in order to avoid breakage. 
- Then add a short comment to all "strange" impls (`Add<&String>`, `Add<&&String>`) referring to the main documentation on `Add<&str>`.

That way we don't have to have duplicated comments/documentation.

Reply:
Okay, I've added a new commit fixing the comments. Please see if what I've revised them with makes sense.